File: topoflow/router.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Discover Legal
"""Macro control loop (spec §9) [AF] + [NEW].

Layered: fold (1) -> typed cells (2) -> UCB1 select (3) -> reward+backup (4) ->
governance/report (5). A topology(dytopo) step is ONE decision-path entry with
ONE backed-up reward, but charges ALL its inner-round tokens.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from . import beliefs as beliefs_mod
from . import governance
from .actions import Plan, legal_actions
from .agents.roles import role_set_for
from .agents.transport import LLMRequest, Transport
from .config import Config
from .policy_graph import PolicyGraph
from .reward.hybrid import hybrid_reward
from .reward.quality import make_quality
from .signature import fold
from .topology.base import CompositeTopologyCell
from .topology.dytopo import DyTopoGenerator
from .topology.linear import LinearWithSkipGenerator
from .trace import (
    Trace,
    build_run_report,
    invoke_event,
    skip_event,
    topology_event,
)
from .types import BeliefVector, CellOutput, HandoffState, TaskContext


@dataclass
class Budget:
    cap: int
    used: int = 0

    def charge(self, t: int) -> None:
        self.used += int(t)

    def exhausted(self) -> bool:
        return self.used >= self.cap


@dataclass
class Trajectory:
    trace: Trace
    reward: float
    quality: float
    sub_scores: dict
    report: object = None
    aborted: bool = False


def make_generator(mode: str, cfg: Config) -> CompositeTopologyCell:
    if mode == "linear":
        return LinearWithSkipGenerator()
    if mode == "dytopo":
        return DyTopoGenerator()
    raise ValueError(f"unknown topology mode {mode!r}")


# ── invoke cells (Layer 2 typed I/O) ────────────────────────────────────────
_SOLVER_HF = "draft_answer"


def run_cell(skill: str, model: str, ctx: TaskContext, handoff: HandoffState,
             transport: Transport, cfg: Config, search_provider=None) -> CellOutput:
    s_pre = skill.lower()
    # web-search cells are tool actions, not LLM calls, when a provider is wired.
    if s_pre.startswith("web_search") and search_provider is not None:
        results = search_provider.search(ctx.prompt, k=5)
        out = CellOutput(skill=skill, model=model, tokens=0)
        out.handoff_delta["evidence"] = results
        return out
    present = [k for k, v in handoff.fields.items() if v is not None]
    req = LLMRequest(
        model=model or cfg.default_model,
        system=f"You are the {skill} cell.",
        user=f"Task: {ctx.prompt}\nKnown: {', '.join(present) or '(nothing)'}",
        schema=("answer",),
        purpose="invoke",
        role=skill,
        meta={"skill": skill},
    )
    fields, tokens = transport.complete(req)
    out = CellOutput(skill=skill, model=model, tokens=tokens, failed=bool(fields.get("_failed")))
    s = skill.lower()
    if s.startswith("solver"):
        ans = fields.get("draft_answer") or fields.get("answer")
        if ans is not None:
            out.handoff_delta[_SOLVER_HF] = ans
    elif s == "planner":
        if fields.get("goal") is not None:
            out.handoff_delta["goal"] = fields["goal"]
        if fields.get("subproblem") is not None:
            out.handoff_delta["subproblem"] = fields["subproblem"]
    elif s == "memory" or s.startswith("web_search"):
        ev = fields.get("evidence")
        if ev is not None:
            out.handoff_delta["evidence"] = ev
    elif s.startswith("verifier"):
        v = fields.get("verification")
        if v is not None:
            out.handoff_delta["verification"] = v
            out.meta["verdict"] = v
    elif s == "evaluator":
        out.complete = bool(fields.get("complete"))
    if out.failed:
        out.retries = 1
    return out


def apply_handoff(handoff: HandoffState, out: CellOutput) -> HandoffState:
    for k, v in out.handoff_delta.items():
        handoff.set(k, v)
    return handoff


def merge_handoff(handoff: HandoffState, merged: HandoffState) -> HandoffState:
    for k, v in merged.fields.items():
        if v is not None:
            handoff.set(k, v)
    return handoff


def evaluator_marks_complete(handoff: HandoffState) -> bool:
    return bool(handoff.get("complete"))


# ── macro loop ──────────────────────────────────────────────────────────────
def run_task(
    ctx: TaskContext,
    cfg: Config,
    policy_graph: PolicyGraph,
    transport: Transport,
    embedder=None,
    quality=None,
    peers: Optional[list] = None,
    select_fn=None,
    search_provider=None,
) -> Trajectory:
    try:
        governance.preflight(ctx, cfg, transport)
    except governance.GovernanceAbort as e:
        trace = Trace(task_id=ctx.task_id)
        report = build_run_report(ctx.task_id, trace, 0.0, 0.0, {"confidence": 1.0}, aborted=True, reason=str(e))
        return Trajectory(trace=trace, reward=0.0, quality=0.0, sub_scores={}, report=report, aborted=True)

    handoff, beliefs = HandoffState(), BeliefVector()
    trace = Trace(task_id=ctx.task_id)
    budget = Budget(cfg.token_cap)
    plan = Plan(list(cfg.optional_aux))
    steps = 0

    while True:
        steps += 1
        sig = fold(ctx, handoff, beliefs, cfg)  # Layer 1
        legal = legal_actions(sig, handoff, plan, budget, cfg)
        if not legal or budget.exhausted() or steps > cfg.max_macro_steps:
            break  # implicit terminate
        # Layer 3, UCB1 (select_fn is an optional policy-injection seam for tests
        # and future policies; it must return one of the legal actions).
        action = select_fn(sig, legal) if select_fn is not None else policy_graph.select(sig, legal)
        trace.record_decision(sig, action)

        if action.kind == "skip":
            plan.mark_skipped(action.target)
            trace.add(skip_event(action))
            continue

        if action.kind == "topology":
            gen = make_generator(action.topo_mode, cfg)
            agents = role_set_for(ctx.domain, cfg)
            res = gen.run(
                ctx, handoff, agents, action, transport, embedder, cfg,
                budget_remaining=budget.cap - budget.used,
            )  # COMPOSITE
            handoff = merge_handoff(handoff, res.merged_handoff)
            beliefs = beliefs_mod.update_beliefs_topology(beliefs, res, cfg)
            budget.charge(res.total_tokens)
            trace.tokens += res.total_tokens
            trace.retries += res.retries
            if res.failed:
                trace.any_failure = True
            trace.add(topology_event(action, res))
            if res.failed or evaluator_marks_complete(handoff):
                break
            continue

        if action.kind == "invoke":
            if action.skill in cfg.optional_aux:
                plan.mark_invoked(action.skill)
            out = run_cell(action.skill, action.model, ctx, handoff, transport, cfg, search_provider)
            handoff = apply_handoff(handoff, out)
            beliefs = beliefs_mod.update_beliefs_cell(beliefs, out, cfg)
            budget.charge(out.tokens)
            trace.tokens += out.tokens
            trace.retries += out.retries
            if out.failed:
                trace.any_failure = True
            trace.add(invoke_event(action, out))
            if action.skill == "evaluator" and out.complete:
                break
            if governance.violation(trace, cfg):
                break

    # final answer for quality scoring
    trace.final_answer = str(handoff.get("merged_answer") or handoff.get("draft_answer") or "")

    q = quality or make_quality(ctx, cfg, transport)
    Q, sub = q.score(ctx, trace, peers or [])
    confidence = float(sub.get("confidence", 1.0))
    r = hybrid_reward(Q, trace.tokens, trace.retries, cfg)

    for (sig_i, action_i) in trace.decision_path():
        policy_graph.backup(sig_i, action_i, r, trace.tokens, trace.any_failure, weight=confidence)

    report = build_run_report(ctx.task_id, trace, r, Q, sub)
    return Trajectory(trace=trace, reward=r, quality=Q, sub_scores=sub, report=report)

File: biglaw-go/internal/topoflow/router.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 Discover Legal

package topoflow

import (
	"errors"
	"fmt"
	"strings"
)

// Budget caps total trajectory tokens.
type Budget struct {
	Cap  int
	Used int
}

func (b *Budget) charge(t int)    { b.Used += t }
func (b *Budget) exhausted() bool { return b.Used >= b.Cap }
func (b *Budget) remaining() int  { return b.Cap - b.Used }

// Trajectory is the outcome of a task run.
type Trajectory struct {
	Trace     *Trace
	Reward    float64
	Quality   float64
	SubScores map[string]any
	Report    *RunReport
	Aborted   bool
}

// SearchProvider is the web-search tool seam (spec §13).
type SearchProvider interface {
	Search(query string, k int) []map[string]any
}

// RunOptions configure a task run.
type RunOptions struct {
	Transport      Transport
	Embedder       Embedder
	Quality        QualitySignal
	Peers          []*Trace
	SearchProvider SearchProvider
	// SelectFn is an optional policy-injection seam (tests / future policies);
	// it must return one of the legal actions. Nil = UCB1 via the policy graph.
	SelectFn func(sig Signature, legal []Action) Action
}

func makeGenerator(mode string) (CompositeTopologyCell, error) {
	switch mode {
	case "linear":
		return LinearWithSkipGenerator{}, nil
	case "dytopo":
		return DyTopoGenerator{}, nil
	}
	return nil, fmt.Errorf("unknown topology mode %q", mode)
}

func runCell(skill, model string, ctx TaskContext, handoff *HandoffState, tx Transport, cfg Config, sp SearchProvider) (CellOutput, error) {
	s := strings.ToLower(skill)
	out := CellOutput{Skill: skill, Model: model, HandoffDelta: map[string]any{}, Meta: map[string]any{}}

	// web-search cells are tool actions, not LLM calls, when a provider is wired.
	if strings.HasPrefix(s, "web_search") && sp != nil {
		res := sp.Search(ctx.Prompt, 5)
		evidence := make([]any, len(res))
		for i, r := range res {
			evidence[i] = r
		}
		out.HandoffDelta["evidence"] = evidence
		return out, nil
	}

	var present []string
	for _, f := range HandoffFields {
		if handoff.Get(f) != nil {
			present = append(present, f)
		}
	}
	fields, tokens, err := tx.Complete(LLMRequest{
		Model: model, System: "You are the " + skill + " cell.",
		User:   "Task: " + ctx.Prompt + "\nKnown: " + strings.Join(present, ", "),
		Schema: []string{"answer"}, Purpose: "invoke", Role: skill, Meta: map[string]any{"skill": skill},
	})
	if err != nil {
		return out, err
	}
	out.Tokens = tokens
	out.Failed = asBool(fields["_failed"])
	switch {
	case strings.HasPrefix(s, "solver"):
		if v := fields["draft_answer"]; v != nil {
			out.HandoffDelta["draft_answer"] = v
		} else if v := fields["answer"]; v != nil {
			out.HandoffDelta["draft_answer"] = v
		}
	case s == "planner":
		if v := fields["goal"]; v != nil {
			out.HandoffDelta["goal"] = v
		}
		if v := fields["subproblem"]; v != nil {
			out.HandoffDelta["subproblem"] = v
		}
	case s == "memory" || strings.HasPrefix(s, "web_search"):
		if v := fields["evidence"]; v != nil {
			out.HandoffDelta["evidence"] = v
		}
	case strings.HasPrefix(s, "verifier"):
		if v := fields["verification"]; v != nil {
			out.HandoffDelta["verification"] = v
			out.Meta["verdict"] = v
		}
	case s == "evaluator":
		out.Complete = asBool(fields["complete"])
	}
	if out.Failed {
		out.Retries = 1
	}
	return out, nil
}

func applyHandoff(h *HandoffState, out CellOutput) {
	for k, v := range out.HandoffDelta {
		h.Set(k, v)
	}
}

func mergeHandoff(h *HandoffState, merged *HandoffState) {
	for k, v := range merged.Fields {
		if v != nil {
			h.Set(k, v)
		}
	}
}

func evaluatorMarksComplete(h *HandoffState) bool { return asBool(h.Get("complete")) }

// RunTask is the macro control loop (spec §9) [AF] + [NEW].
func RunTask(ctx TaskContext, cfg Config, pg *PolicyGraph, opts RunOptions) (*Trajectory, error) {
	tx := opts.Transport
	if err := preflight(ctx, cfg, tx); err != nil {
		if errors.Is(err, ErrGovernanceAbort) {
			tr := NewTrace(ctx.TaskID)
			rep := buildRunReport(tr, 0, 0, map[string]any{"confidence": 1.0}, true, err.Error())
			return &Trajectory{Trace: tr, Report: rep, Aborted: true}, nil
		}
		return nil, err
	}

	handoff := NewHandoffState()
	beliefs := NewBeliefVector()
	trace := NewTrace(ctx.TaskID)
	budget := &Budget{Cap: cfg.TokenCap}
	plan := NewPlan(cfg.OptionalAux)
	steps := 0

	for {
		steps++
		sig := Fold(ctx, handoff, beliefs, cfg)
		legal := legalActions(plan, cfg)
		if len(legal) == 0 || budget.exhausted() || steps > cfg.MaxMacroSteps {
			break
		}
		var action Action
		if opts.SelectFn != nil {
			action = opts.SelectFn(sig, legal)
		} else {
			action = pg.Select(sig, legal)
		}
		trace.recordDecision(sig, action)

		switch action.Kind {
		case "skip":
			plan.markSkipped(action.Target)
			trace.add(skipEvent(action))
			continue

		case "topology":
			gen, err := makeGenerator(action.TopoMode)
			if err != nil {
				return nil, err
			}
			agents := RoleSetFor(ctx.Domain, cfg)
			res, err := gen.Run(ctx, handoff, agents, action, tx, opts.Embedder, cfg, budget.remaining())
			if err != nil {
				return nil, err
			}
			mergeHandoff(handoff, res.MergedHandoff)
			beliefs = updateBeliefsTopology(beliefs, res)
			budget.charge(res.TotalTokens)
			trace.Tokens += res.TotalTokens
			trace.Retries += res.Retries
			if res.Failed {
				trace.AnyFailure = true
			}
			trace.add(topologyEvent(action, res))
			if res.Failed || evaluatorMarksComplete(handoff) {
				goto finish
			}
			continue

		case "invoke":
			if cfg.IsOptionalAux(action.Skill) {
				plan.markInvoked(action.Skill)
			}
			out, err := runCell(action.Skill, action.Model, ctx, handoff, tx, cfg, opts.SearchProvider)
			if err != nil {
				return nil, err
			}
			applyHandoff(handoff, out)
			beliefs = updateBeliefsCell(beliefs, out)
			budget.charge(out.Tokens)
			trace.Tokens += out.Tokens
			trace.Retries += out.Retries
			if out.Failed {
				trace.AnyFailure = true
			}
			trace.add(invokeEvent(action, out))
			if action.Skill == "evaluator" && out.Complete {
				goto finish
			}
			if violation(trace, cfg) {
				goto finish
			}
		}
	}

finish:
	if v := handoff.Get("merged_answer"); v != nil {
		trace.FinalAnswer = fmt.Sprintf("%v", v)
	} else if v := handoff.Get("draft_answer"); v != nil {
		trace.FinalAnswer = fmt.Sprintf("%v", v)
	}

	q := opts.Quality
	if q == nil {
		q = MakeQuality(ctx, cfg, nil, nil)
	}
	Q, sub := q.Score(ctx, trace, opts.Peers)
	confidence := 1.0
	if c, ok := asFloat(sub["confidence"]); ok {
		confidence = c
	}
	r := hybridReward(Q, trace.Tokens, trace.Retries, cfg)

	for _, d := range trace.DecisionPath() {
		pg.Backup(d.Sig, d.Act, r, trace.Tokens, trace.AnyFailure, confidence)
	}

	report := buildRunReport(trace, r, Q, sub, false, "")
	return &Trajectory{Trace: trace, Reward: r, Quality: Q, SubScores: sub, Report: report}, nil
}
